# W09 - Association Rule
In this hands-on session, we will try to uncover the association rule from given transaction items. Association rule mining is a technique to identify underlying relations between different items. We will cover:
1. data preprocessing for association rule,
2. implement algorithm for association rule,
3. find interesting association rule from dataset

## Read dataset "dataset_association_rule.dat"
Download the dataset "dataset_association_rule.dat" in the course portal "IF4041 Ilmu Data dan Penggalian Data".

In [1]:
import numpy as np
import pandas as pd

df = pd.read_csv('dataset_association_rule.dat', header=None)
df.head()

0
0  36 69 115 226 278 343 345 358 368 370 401 450 ...
1  8 51 55 73 78 117 140 175 187 229 266 295 304 ...
2  53 55 98 159 192 322 332 402 412 413 424 430 4...
3  48 203 205 227 279 294 320 335 359 414 509 525...
4  58 69 73 82 120 146 166 178 217 220 244 256 33...

## Question 01 (Q01)
Based on dataset that given to you, implement several data pre-processing process. <b>Note :</b> you should taylor the data pre-processing step based on algorithm and library that you used in this hands-on. Save the result of data pre-processing in variable called 'item_belanja'.<br>

In [16]:
dataset = df.values.tolist()
for row in dataset:
    item_belanja.append(row[0].split())

Print the pre-processing result :

In [31]:
item_belanja

[['36',
  '69',
  '115',
  '226',
  '278',
  '343',
  '345',
  '358',
  '368',
  '370',
  '401',
  '450',
  '489',
  '494',
  '573',
  '577',
  '581',
  '583',
  '610',
  '682',
  '692',
  '705',
  '722',
  '832',
  '862',
  '886',
  '908',
  '923',
  '932',
  '960',
  '977'],
 ['8',
  '51',
  '55',
  '73',
  '78',
  '117',
  '140',
  '175',
  '187',
  '229',
  '266',
  '295',
  '304',
  '366',
  '381',
  '413',
  '424',
  '429',
  '501',
  '512',
  '523',
  '529',
  '538',
  '572',
  '575',
  '576',
  '593',
  '675',
  '676',
  '688',
  '735',
  '758',
  '785',
  '797',
  '812',
  '823',
  '826',
  '843',
  '854',
  '868',
  '871',
  '888',
  '893',
  '956',
  '982'],
 ['53',
  '55',
  '98',
  '159',
  '192',
  '322',
  '332',
  '402',
  '412',
  '413',
  '424',
  '430',
  '450',
  '480',
  '526',
  '538',
  '569',
  '571',
  '572',
  '598',
  '666',
  '672',
  '694',
  '701',
  '797',
  '809',
  '820',
  '826',
  '897',
  '904',
  '928',
  '943',
  '952',
  '956',
  '992'],
 ['48',
 

Praproses data menggunakan `TransactionEncoder`

In [33]:
from mlxtend.frequent_patterns import apriori

te_ary = te.fit(item_belanja).transform(item_belanja)
preprocessed_df = pd.DataFrame(te_ary, columns=te.columns_)
preprocessed_df

0      1     10    100    101    102    103    104    105    106  \
0       False  False  False  False  False  False  False  False  False  False   
1       False  False  False  False  False  False  False  False  False  False   
2       False  False  False  False  False  False  False  False  False  False   
3       False  False  False  False  False  False  False  False  False  False   
4       False  False  False  False  False  False  False  False  False  False   
...       ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
299995  False  False  False  False  False  False  False  False  False  False   
299996  False  False  False  False  False  False  False  False  False  False   
299997  False  False  False  False  False  False  False  False  False  False   
299998  False  False  False  False  False  False  False  False  False  False   
299999  False  False  False  False  False  False  False  False  False  False   

        ...    989    990    991    992    993    994    995    997    998  \
0       ...  False  False  False  False  False  False  False  False  False   
1       ...  False  False  False  False  False  False  False  False  False   
2       ...  False  False  False   True  False  False  False  False  False   
3       ...  False  False  False  False  False  False  False  False  False   
4       ...  False  False  False  False  False  False  False  False  False   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
299995  ...   True  False  False  False  False  False  False  False  False   
299996  ...  False  False  False  False  False  False  False  False  False   
299997  ...  False  False  False  False  False  False  False  False  False   
299998  ...  False  False  False  False  False  False  False  False  False   
299999  ...  False  False  False  False  False  False  False  False  False   

          999  
0       False  
1       False  
2       False  
3       False  
4       False  
...       ...  
299995  False  
299996  False  
299997  False  
299998  False  
299999  False  

[300000 rows x 942 columns]

## Question 02 (Q02)
In this step, you will generate association rule from this dataset.<br><br>
a. Choose the library that you will use for generating association rule, and give explanation about why you choose that library and what is the algorithm used in that library<br>
b. Using the library that you've chosen in (Q02.a), generate association rules from the pre-processed dataset in Q01.<br>
c. Print the association rule that you have find in Q02.b along with its support and confidence.<br>
d. What is your conclusion from association rule that you have uncover before ?

Your Answer (Q02.a):

Association rule generation menggunakan library [**MLextend**](http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/#frequent-itemsets-via-apriori-algorithm)
karena library yang disediakan beserta dokumentasinya lengkap, terstruktur, dan mudah dipelajari.

Pembangkitan frequent itemsets menggunakan fpgrowth

In [56]:
from mlxtend.frequent_patterns import fpgrowth
from mlxtend.frequent_patterns import association_rules

frequent_itemsets = fpgrowth(preprocessed_df, min_support=0.01)
frequent_itemsets

support    itemsets
0      0.28738       (282)
1      0.18921       (407)
2      0.17427       (610)
3      0.17203       (621)
4      0.16203       (654)
...        ...         ...
65231  0.01184   (25, 852)
65232  0.01381  (852, 878)
65233  0.01091  (852, 126)
65234  0.01043  (329, 171)
65235  0.01041  (395, 431)

[65236 rows x 2 columns]

In [60]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.3)
rules

antecedents consequents  antecedent support  consequent support  \
0             (407)       (282)             0.18921             0.28738   
1        (126, 407)       (282)             0.03695             0.28738   
2        (126, 407)       (450)             0.03695             0.23384   
3        (336, 407)       (282)             0.03347             0.28738   
4        (763, 407)       (282)             0.03915             0.28738   
...             ...         ...                 ...                 ...   
3941189       (533)       (450)             0.01558             0.23384   
3941190       (852)       (398)             0.04360             0.13540   
3941191       (852)       (878)             0.04360             0.16311   
3941192       (171)       (329)             0.01165             0.06191   
3941193       (395)       (431)             0.02525             0.20125   

         support  confidence       lift  leverage  conviction  
0        0.06120    0.323450   1.125514  0.006825    1.053315  
1        0.01530    0.414073   1.440856  0.004681    1.216227  
2        0.01201    0.325034   1.389984  0.003370    1.135109  
3        0.01096    0.327457   1.139458  0.001341    1.059591  
4        0.01289    0.329246   1.145683  0.001639    1.062417  
...          ...         ...        ...       ...         ...  
3941189  0.01011    0.648909   2.775012  0.006467    2.182225  
3941190  0.01352    0.310092   2.290190  0.007617    1.253210  
3941191  0.01381    0.316743   1.941899  0.006698    1.224854  
3941192  0.01043    0.895279  14.460975  0.009709    8.957991  
3941193  0.01041    0.412277   2.048582  0.005328    1.359059  

[3941194 rows x 9 columns]

In [65]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.99)
rules

antecedents consequents  antecedent support  \
0               (450, 259, 636, 330)       (146)             0.01033   
1          (450, 259, 228, 330, 636)       (146)             0.01011   
2          (737, 450, 259, 228, 330)       (146)             0.01014   
3               (737, 330, 259, 636)       (146)             0.01037   
4               (636, 737, 259, 228)       (146)             0.01035   
...                              ...         ...                 ...   
20222       (503, 209, 83, 407, 185)       (741)             0.01253   
20223           (641, 185, 209, 503)       (741)             0.01276   
20224       (641, 209, 83, 503, 185)       (741)             0.01259   
20225      (641, 503, 209, 407, 185)       (741)             0.01250   
20226  (641, 503, 209, 83, 407, 185)       (741)             0.01233   

       consequent support  support  confidence      lift  leverage  conviction  
0                 0.11040  0.01027    0.994192  9.005359  0.009130  153.159467  
1                 0.11040  0.01008    0.997033  9.031093  0.008964  299.795200  
2                 0.11040  0.01009    0.995069  9.013306  0.008971  180.410880  
3                 0.11040  0.01028    0.991321  8.979358  0.009135  102.501689  
4                 0.11040  0.01030    0.995169  9.014213  0.009157  184.147200  
...                   ...      ...         ...       ...       ...         ...  
20222             0.10072  0.01241    0.990423  9.833429  0.011148   93.899820  
20223             0.10072  0.01264    0.990596  9.835143  0.011355   95.623440  
20224             0.10072  0.01247    0.990469  9.833882  0.011202   94.349460  
20225             0.10072  0.01238    0.990400  9.833201  0.011121   93.675000  
20226             0.10072  0.01221    0.990268  9.831887  0.010968   92.401020  

[20227 rows x 9 columns]

Your Conclusion (Q02.d):

Dari percobaan men-generate frequent_itemsets dengan minimum support **0.01** dengan minimum threshold **0.99**, dihasilkan sebanyak 20227 rule.

Pada frequent itemsets, semakin kecil nilai minimum support, semakin banyak frequent itemset yang dihasilkan.

Pada association rules, semakin kecil nilai minimum threshold, semakin sedikit rules yang dihasilkan.